// @0x/utils/lib/src/abi_encoder/evm_data_type_factory.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var signature_parser_1 = require("./utils/signature_parser");
var address_1 = require("./evm_data_types/address");
var array_1 = require("./evm_data_types/array");
var bool_1 = require("./evm_data_types/bool");
var dynamic_bytes_1 = require("./evm_data_types/dynamic_bytes");
var int_1 = require("./evm_data_types/int");
var method_1 = require("./evm_data_types/method");
var pointer_1 = require("./evm_data_types/pointer");
var static_bytes_1 = require("./evm_data_types/static_bytes");
var string_1 = require("./evm_data_types/string");
var tuple_1 = require("./evm_data_types/tuple");
var uint_1 = require("./evm_data_types/uint");
var Address = /** @class */ (function (_super) {
    __extends(Address, _super);
    function Address(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return Address;
}(address_1.AddressDataType));
exports.Address = Address;
var Bool = /** @class */ (function (_super) {
    __extends(Bool, _super);
    function Bool(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return Bool;
}(bool_1.BoolDataType));
exports.Bool = Bool;
var Int = /** @class */ (function (_super) {
    __extends(Int, _super);
    function Int(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return Int;
}(int_1.IntDataType));
exports.Int = Int;
var UInt = /** @class */ (function (_super) {
    __extends(UInt, _super);
    function UInt(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return UInt;
}(uint_1.UIntDataType));
exports.UInt = UInt;
var StaticBytes = /** @class */ (function (_super) {
    __extends(StaticBytes, _super);
    function StaticBytes(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return StaticBytes;
}(static_bytes_1.StaticBytesDataType));
exports.StaticBytes = StaticBytes;
var DynamicBytes = /** @class */ (function (_super) {
    __extends(DynamicBytes, _super);
    function DynamicBytes(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return DynamicBytes;
}(dynamic_bytes_1.DynamicBytesDataType));
exports.DynamicBytes = DynamicBytes;
var String = /** @class */ (function (_super) {
    __extends(String, _super);
    function String(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return String;
}(string_1.StringDataType));
exports.String = String;
var Pointer = /** @class */ (function (_super) {
    __extends(Pointer, _super);
    function Pointer(destDataType, parentDataType) {
        return _super.call(this, destDataType, parentDataType, EvmDataTypeFactory.getInstance()) || this;
    }
    return Pointer;
}(pointer_1.PointerDataType));
exports.Pointer = Pointer;
var Tuple = /** @class */ (function (_super) {
    __extends(Tuple, _super);
    function Tuple(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return Tuple;
}(tuple_1.TupleDataType));
exports.Tuple = Tuple;
var Array = /** @class */ (function (_super) {
    __extends(Array, _super);
    function Array(dataItem) {
        return _super.call(this, dataItem, EvmDataTypeFactory.getInstance()) || this;
    }
    return Array;
}(array_1.ArrayDataType));
exports.Array = Array;
var Method = /** @class */ (function (_super) {
    __extends(Method, _super);
    function Method(abi) {
        return _super.call(this, abi, EvmDataTypeFactory.getInstance()) || this;
    }
    return Method;
}(method_1.MethodDataType));
exports.Method = Method;
/* tslint:disable no-construct */
var EvmDataTypeFactory = /** @class */ (function () {
    /* tslint:enable prefer-function-over-method */
    function EvmDataTypeFactory() {
    }
    EvmDataTypeFactory.getInstance = function () {
        if (!EvmDataTypeFactory._instance) {
            EvmDataTypeFactory._instance = new EvmDataTypeFactory();
        }
        return EvmDataTypeFactory._instance;
    };
    /* tslint:disable prefer-function-over-method */
    EvmDataTypeFactory.prototype.create = function (dataItem, parentDataType) {
        // Create data type
        var dataType;
        if (Array.matchType(dataItem.type)) {
            dataType = new Array(dataItem);
        }
        else if (Address.matchType(dataItem.type)) {
            dataType = new Address(dataItem);
        }
        else if (Bool.matchType(dataItem.type)) {
            dataType = new Bool(dataItem);
        }
        else if (Int.matchType(dataItem.type)) {
            dataType = new Int(dataItem);
        }
        else if (UInt.matchType(dataItem.type)) {
            dataType = new UInt(dataItem);
        }
        else if (StaticBytes.matchType(dataItem.type)) {
            dataType = new StaticBytes(dataItem);
        }
        else if (Tuple.matchType(dataItem.type)) {
            dataType = new Tuple(dataItem);
        }
        else if (DynamicBytes.matchType(dataItem.type)) {
            dataType = new DynamicBytes(dataItem);
        }
        else if (String.matchType(dataItem.type)) {
            dataType = new String(dataItem);
        }
        // @TODO: DataTypeement Fixed/UFixed types
        if (_.isUndefined(dataType)) {
            throw new Error("Unrecognized data type: '" + dataItem.type + "'");
        }
        else if (!_.isUndefined(parentDataType) && !dataType.isStatic()) {
            var pointerToDataType = new Pointer(dataType, parentDataType);
            return pointerToDataType;
        }
        return dataType;
    };
    return EvmDataTypeFactory;
}());
exports.EvmDataTypeFactory = EvmDataTypeFactory;
/**
 * Convenience function for creating a DataType from different inputs.
 * @param input A single or set of DataItem or a DataType signature.
 *              A signature in the form of '<type>' is interpreted as a `DataItem`
 *              For example, 'string' is interpreted as {type: 'string'}
 *              A signature in the form '(<type1>, <type2>, ..., <typen>)' is interpreted as `DataItem[]`
 *              For eaxmple, '(string, uint256)' is interpreted as [{type: 'string'}, {type: 'uint256'}]
 * @return DataType corresponding to input.
 */
function create(input) {
    // Handle different types of input
    var isSignature = typeof input === 'string';
    var isTupleSignature = isSignature && input.startsWith('(');
    var shouldParseAsTuple = isTupleSignature || _.isArray(input);
    // Create input `dataItem`
    var dataItem;
    if (shouldParseAsTuple) {
        var dataItems = isSignature ? signature_parser_1.generateDataItemsFromSignature(input) : input;
        dataItem = {
            name: '',
            type: 'tuple',
            components: dataItems,
        };
    }
    else {
        dataItem = isSignature ? signature_parser_1.generateDataItemsFromSignature(input)[0] : input;
    }
    // Create data type
    var dataType = EvmDataTypeFactory.getInstance().create(dataItem);
    return dataType;
}
exports.create = create;
/* tslint:enable no-construct */
